migrations/entitlements/migration.go

```diff
@@ -0,0 +1,245 @@
+	case *sema.VariableSizedType:
+		return sema.NewVariableSizedType(nil, ConvertToEntitledType(t.Type))
+	case *sema.ConstantSizedType:
+		return sema.NewConstantSizedType(nil, ConvertToEntitledType(t.Type), t.Size)
+	case *sema.DictionaryType:
+		return sema.NewDictionaryType(nil, ConvertToEntitledType(t.KeyType), ConvertToEntitledType(t.ValueType))
+	default:
+		return t
+	}
+}
+
+// Converts the input value into a version compatible with the new entitlements feature,
+// with the same members/operations accessible on any references as would have been accessible in the past.
+func ConvertValueToEntitlements(
+	inter *interpreter.Interpreter,
+	v interpreter.Value,
+) interpreter.Value {
+
+	var staticType interpreter.StaticType
+	// for reference types, we want to use the borrow type, rather than the type of the referenced value
+	switch referenceValue := v.(type) {
+	case *interpreter.EphemeralReferenceValue:
+		staticType = interpreter.NewReferenceStaticType(
+			inter,
+			referenceValue.Authorization,
+			interpreter.ConvertSemaToStaticType(inter, referenceValue.BorrowedType),
+		)
+	case *interpreter.StorageReferenceValue:
+		staticType = interpreter.NewReferenceStaticType(
+			inter,
+			referenceValue.Authorization,
+			interpreter.ConvertSemaToStaticType(inter, referenceValue.BorrowedType),
+		)
+	default:
+		staticType = v.StaticType(inter)
+	}
+
+	// if the static type contains a legacy restricted type, convert it to a new type according to some rules:
+	// &T{I} -> auth(SupportedEntitlements(I)) &T
+	// Capability<&T{I}> -> Capability<auth(SupportedEntitlements(I)) &T>
+	var convertLegacyStaticType func(interpreter.StaticType)
+	convertLegacyStaticType = func(staticType interpreter.StaticType) {
+		switch t := staticType.(type) {
+		case *interpreter.ReferenceStaticType:
+			switch referencedType := t.ReferencedType.(type) {
+			case *interpreter.IntersectionStaticType:
+				if referencedType.LegacyType != nil {
+					t.ReferencedType = referencedType.LegacyType
+					intersectionSemaType := inter.MustConvertStaticToSemaType(referencedType).(*sema.IntersectionType)
+					auth := sema.UnauthorizedAccess
+					supportedEntitlements := intersectionSemaType.SupportedEntitlements()
+					if supportedEntitlements.Len() > 0 {
+						auth = sema.EntitlementSetAccess{
+							SetKind:      sema.Conjunction,
+							Entitlements: supportedEntitlements,
+						}
+					}
+					t.Authorization = interpreter.ConvertSemaAccessToStaticAuthorization(inter, auth)
+				}
+			}
+		case *interpreter.CapabilityStaticType:
+			convertLegacyStaticType(t.BorrowType)
+		case *interpreter.VariableSizedStaticType:
+			convertLegacyStaticType(t.Type)
+		case *interpreter.ConstantSizedStaticType:
+			convertLegacyStaticType(t.Type)
+		case *interpreter.DictionaryStaticType:
+			convertLegacyStaticType(t.KeyType)
+			convertLegacyStaticType(t.ValueType)
+		case *interpreter.OptionalStaticType:
+			convertLegacyStaticType(t.Type)
+		}
```

Comment:
Maybe also convert `LegacyType` of `IntersectionStaticType`

Reply:
This case is impossible; the instersection only has interfaces in it, and the legacy type of the `IntersectionStaticType` is only ever a composite